alf/experience_replayers/replay_buffer.py

```diff
@@ -518,49 +523,50 @@ def _pad(self, x, env_ids):
-            discounted_return[
-                mask] = r + self._gamma * disc * discounted_return[mask]
-            # point to previous step
-            current_pos -= 1
-            # make sure it's not underflowing
-            valid = current_pos >= self._current_size[
-                env_ids] - self._max_length
-            ind = (env_ids[valid], self.circular(current_pos[valid]))
-            # save to buffer
-            if torch.any(valid):
-                self._episodic_discounted_return[ind] = discounted_return[
-                    mask][valid]
-            # remove env_ids that are already done
-            step_type = self._buffer.step_type[(env_ids,
-                                                self.circular(current_pos))]
-            valid &= step_type != ds.StepType.FIRST
-            env_ids, current_pos = env_ids[valid], current_pos[valid]
-            mask[mask.clone()] &= valid
+        ind = (env_ids, self.circular(self._current_pos[env_ids] - 1))
+        self._episodic_discounted_return[ind] = self._default_return
+
+    def _compute_store_episodic_return(self, env_ids):
+        # Always pass in env_ids whose discount is 0, to save computation.
+        current_pos = self._current_pos[env_ids]
+        current_pos -= 1
+
+        # Store episodic return when discount == 0, not ending with timelimit.
+        # TODO: bootstrapping the the episodic return with the default_return
+        # if game ends with timelimit. For example, the episode length of
+        # Atari games is limited to 4500, which can be reached quite often.
+        first_step_pos = self.get_episode_begin_position(current_pos, env_ids)
+        headless = first_step_pos < current_pos - self._max_length + 1
+        first_step_pos[headless] = (
+            current_pos - self._max_length + 1)[headless]
+        max_len = torch.max(current_pos - first_step_pos) + 1
+        # Indexes cover all env_ids for the length of the longest episode.
+        epi_pos = current_pos.unsqueeze(1) - max_len + 1 + torch.arange(
+            max_len).unsqueeze(0)
+        all_ind = (env_ids.unsqueeze(1), self.circular(epi_pos))
+        # mask for the real indices
+        valid = epi_pos >= first_step_pos.unsqueeze(1)
+        # For episode with [FIRST, MID, MID, LAST] steps,
+        # discounts = [1, gamma, gamma^2, gamma^3]
+        discounts = (self._gamma
+                     **torch.arange(max_len, dtype=torch.float32)).unsqueeze(0)
+        epi_discounts = discounts * self._buffer.discount[all_ind]
+        # rewards = [r0, r1, r2, r3]
+        rewards = self._buffer.reward[all_ind]
+        rewards[~valid] = 0
+        epi_discounts[~valid] = 0
+        # reward[t] * discount[t - 1] is the correct discounted_return.
+        discounted_rewards = rewards[:, 1:] * epi_discounts[:, :-1]
+        # Then flip and cumsum from back of episode.
+        discounted_return = discounted_rewards.flip(1).cumsum(dim=1).flip(1)
```

Comment:
There is a problem with this if some TimeStep.discount is 0 in the middle of the episode. cumsum will add the reward after the zero discount towards the steps before it.
Is this tested in unittest?